File: Introduction_to_Software_Testing/C/my_sqrt.c
#include <stdio.h>
#include <assert.h>
#include <math.h>
#include <stdlib.h>
#include <time.h>
#include <string.h> 

#define EPSILON 1e-8

double my_sqrt(double x) {
    double value = 0; 
    double guess = x / 2; 
    while(value != guess) {
        value = guess; 
        guess = ( value + x / value )/ 2; 
    }
    return value; 
}

double my_sqrt_with_log(double x) {
    double value = 0; 
    double guess = x / 2; 
    while(value != guess) {
        printf("%.6f\n", value); 
        value = guess; 
        guess = ( value + x / value )/ 2; 
    }
    return value; 
}

void assertEquals(double x,double y) {
    assert(fabs(x-y) < EPSILON);
}

double my_sqrt_checked(double x) { 
    double root = my_sqrt(x) ; 
    assertEquals(root * root, x); 
    return root ; 
}

void sqrt_program(char* arg) {
    int x = atoi(arg); 
    
    if( x == 0) {
        printf("Illegal Input\n");
    }else{
        if (x < 0) {
            printf("Illegal Number\n"); 
        }else{
            printf("The root of %d is %.6lf\n", x, my_sqrt(x));
        }
    }
}

double my_sqrt_fixed(int x) {
    assert(0 <= x); 
    if(x == 0) {
        return 0 ; 
    }
    return my_sqrt(x) ;
}

int main(void) {
    int x = 0; 
    scanf("%d", &x) ;

    printf("%.6lf\n", my_sqrt(x)); 

    int result = my_sqrt(x) ;
    float expected_result = 2.0 ; 
    if(result == expected_result) {
        printf("Test passed\n");
    }else{
        printf("Test failed"); 
    }

    printf("ASSERT TEST\n"); 
    assert(my_sqrt(4) == 2) ;
    
    assert(fabs(my_sqrt(4) - 2) < EPSILON); 

    // Test pass
    assertEquals(my_sqrt(4), 2); 
    assertEquals(my_sqrt(9), 3); 
    assertEquals(my_sqrt(100), 10); 

    // Generating Test 
    assertEquals(my_sqrt(2) * my_sqrt(2), 2); 
    assertEquals(my_sqrt(3) * my_sqrt(3), 3);
    assertEquals(my_sqrt(42.11) * my_sqrt(42.11), 42.11);

    for(int i = 1; i < 1000; i++) { 
        assertEquals(my_sqrt(i) * my_sqrt(i), i);
    }

    // Timer 
    clock_t start, end ; 
    start = clock(); 
    for(int i = 0 ; i < 10000; i++) {
        assertEquals(my_sqrt(i) * my_sqrt(i), i) ; 
    }
    end = clock(); 
    printf("WORKING TIME: %.6lf\n", (double)(end - start)/CLOCKS_PER_SEC);
    
    // Random
    srand(time(NULL));

    start = clock(); 
    for(int i = 0 ; i < 10000; i++) {
        int x = 1 + rand() * 1000000; 
        assertEquals(my_sqrt(i) * my_sqrt(i), i) ; 
    }
    end = clock(); 
    printf("WORKING TIME: %.6lf\n", (double)(end - start)/CLOCKS_PER_SEC);
    
    printf("%.6lf\n", my_sqrt_checked(2.0)); 

    sqrt_program("4"); 
    sqrt_program("-1");
    sqrt_program("xyzzy");
    
    return 0; 
}

File: Introduction_to_Software_Testing/Python/my_sqrt.py

from timer import timer
import random
import logging
# from ExpectError import ExpectError
# from ExpectError ipmort ExpectError

def my_sqrt(x):
    """Computes the square root of x, using the Newton-Raphson method"""
    approx = None
    guess = x / 2
    while approx != guess:
        approx = guess
        guess = (approx + x / approx) / 2
    return approx

def my_sqrt_with_log(x):
    """Computes the square root of x, using the Newton-Raphson method"""
    approx = 0
    guess = x / 2
    while approx != guess:
        print("approx ="+str(approx))
        approx = guess
        guess = (approx + x / approx) / 2
    return approx  

def assertEquals(x, y, epsilon=1e-8):
    assert abs(x - y) < epsilon


def my_sqrt_checked(x): 
    root = my_sqrt(x) 
    assertEquals(root * root, x) 
    return root 

def sqrt_program(arg):
    try:
        x = int(arg) 
    except ValueError:
        print("Illegal Input")
    else:
        if x < 0:
            print("Illegal Number") 
        else:
            print('The root of', x, 'is', my_sqrt(x))

def my_sqrt_fixed(x):
    assert 0 <= x
    if x == 0:
        return 0
    return my_sqrt(x)

x = int(input())

##
# print(my_sqrt_with_log(x))
##

## Automating Test Execution
# issue:
# 1. we need five lines of code for a single test
# 2. we do not care for rounding errors 
# 3. we only check a single input (and a signle result)
##
result = my_sqrt(x) 
expected_result = 2.0
if result == expected_result: 
    print("Test passed") 
else:
    print("Test failed")

##
print("ASSERT TEST")
assert my_sqrt(4) == 2

EPSILON = 1e-8
assert abs(my_sqrt(4) - 2) < EPSILON
##

## Test pass 
assertEquals(my_sqrt(4), 2)
assertEquals(my_sqrt(9), 3)
assertEquals(my_sqrt(100), 10)

## Generating Test 
assertEquals(my_sqrt(2) * my_sqrt(2), 2)
assertEquals(my_sqrt(3) * my_sqrt(3), 3)
assertEquals(my_sqrt(42.11) * my_sqrt(42.11), 42.11)

for n in range(1, 1000):
    assertEquals(my_sqrt(n) * my_sqrt(n), n) 


## Timer 
logging.basicConfig(level=logging.DEBUG)
with timer() as t:
    for n in range(1, 10000):
        assertEquals(my_sqrt(n) * my_sqrt(n), n)
# print(t.elapsed_time())

## Random 
logging.basicConfig(level=logging.DEBUG)
with timer() as t:
    for i in range(10000):
        x = 1 + random.random() * 1000000
        assertEquals(my_sqrt(x) * my_sqrt(x), x) 
# print(t.elapsed_time())

print(my_sqrt_checked(2.0))

sqrt_program("4")

# 모듈 없음 
# with ExpectTimeout(1):
sqrt_program("-1")

sqrt_program("xyzzy")

assert my_sqrt_fixed(0) == 0
assert my_sqrt_fixed(-1) == 0

# our best chance with testing is to 
# 1. Test the program on several, well-chosen inputs; and 
# 2. Check results extensively and automatically 
# This is what we do in the remainder of this course: 
# Devise techniques that help us to thoroughly test a program, as well as techniques that help us checking its state for correctness.

